DigitalLearningSolutions.Web/Controllers/MyAccountController.cs

```diff
@@ -49,23 +56,87 @@ public IActionResult EditDetails()
         }
 
         [HttpPost]
-        public IActionResult EditDetails(EditDetailsViewModel model)
+        public IActionResult EditDetails(EditDetailsViewModel model, string action)
+        {
+            switch (action)
+            {
+                case "save":
+                    return EditDetailsPostSave(model);
+                case "previewImage":
+                    return EditDetailsPostPreviewImage(model);
+                case "removeImage":
+                    return EditDetailsPostRemoveImage(model);
+                default:
+                    return View(model);
+            }
+        }
+
+        private IActionResult EditDetailsPostSave(EditDetailsViewModel model)
         {
             if (!ModelState.IsValid)
             {
                 return View(model);
             }
 
+            if (model.ProfilePicture != null)
+            {
+                ModelState.AddModelError("ProfilePicture", "Preview your new profile picture before saving");
+                return View(model);
+            }
+
             var userAdminId = User.GetAdminId();
             var userDelegateId = User.GetNullableCandidateId();
 
-            if (!userService.TryUpdateUserAccountDetails(userAdminId, userDelegateId, model.Password,  model.FirstName, model.LastName, model.Email))
+            var profileImageToSave = model.HasProfileImageBeenRemoved ? null : model.NewProfileImage ?? model.CurrentProfileImage;
+
+            if (!userService.TryUpdateUserAccountDetails(userAdminId, userDelegateId, model.Password,  model.FirstName, model.LastName, model.Email, profileImageToSave))
             {
                 ModelState.AddModelError("Password", "The password you have entered is incorrect.");
                 return View(model);
             }
 
             return RedirectToAction("Index");
         }
+
+        private IActionResult EditDetailsPostPreviewImage(EditDetailsViewModel model)
+        {
+            // We don't want to display validation errors on other fields in this case
+            foreach (var key in ModelState.Keys.Where(k => k != "ProfilePicture"))
+            {
+                ModelState[key].Errors.Clear();
+                ModelState[key].ValidationState = ModelValidationState.Valid;
+            }
+
+            if (!ModelState.IsValid)
+            {
+                return View(model);
+            }

```

Comment:
We've just ensured the ModelState is valid so why are we checking that here?

Reply:
We've just cleared all errors except errors on the ProfilePicture property. This then throws it back out if the ProfilePicture is invalid (e.g. wrong file type)

Reply:
Oh yeah, I just totally ignored the `k != "ProfilePicture"` bit 🤦‍♂️